updates_and_signals/safe_message_handlers/src/cluster-manager.ts

```diff
@@ -0,0 +1,134 @@
+      if (!new Set(this.state.nodes.values()).has(input.jobName)) {
+        const unassignedNodes = this.getUnassignedNodes();
+        if (input.numNodes > unassignedNodes.size) {
+          throw new wf.ApplicationFailure(
+            `Cannot assign ${input.numNodes} nodes; have only ${unassignedNodes.size} available`
+          );
+        }
+        const nodesToAssign = Array.from(unassignedNodes).slice(0, input.numNodes);
+        // This await would be dangerous without the lock held because it would allow interleaving
+        // with the deleteJob operation, which mutates self.state.nodes.
+        await assignNodesToJob({ nodes: nodesToAssign, jobName: input.jobName });
+        for (const node of nodesToAssign) {
+          this.state.nodes.set(node, input.jobName);
+        }
+        this.state.maxAssignedNodes = Math.max(this.state.maxAssignedNodes, this.getAssignedNodes().size);
+      }
+      return this.getStateSummary();
+    });
+  }
+
+  async deleteJob(input: DeleteJobUpdateInput) {
+    await wf.condition(() => this.state.clusterStarted);
+    if (this.state.clusterShutdown) {
+      // If you want the client to receive a failure, either add an update validator and throw the
+      // exception from there, or raise an ApplicationError. Other exceptions in the handler will
+      // cause the workflow to keep retrying and get it stuck.
+      throw new wf.ApplicationFailure('Cannot delete job: Cluster is already shut down');
+    }
+    await this.nodesMutex.runExclusive(async () => {
+      const nodesToUnassign = Array.from(this.state.nodes.entries())
+        .filter(([_, v]) => v === input.jobName)
+        .map(([k, _]) => k);
+      // This await would be dangerous without the lock held because it would allow interleaving
+      // with the assignNodesToJob and performHealthCheck operations, both of which mutate
+      // self.state.nodes.
+      await unassignNodesForJob({ nodes: nodesToUnassign, jobName: input.jobName });
+      for (const node of nodesToUnassign) {
+        this.state.nodes.set(node, null);
+      }
+    });
+  }
+
+  getState(): ClusterManagerState {
+    return {
+      clusterStarted: this.state.clusterStarted,
+      clusterShutdown: this.state.clusterShutdown,
+      nodes: this.state.nodes,
+      maxAssignedNodes: this.state.maxAssignedNodes,
+    };
+  }
+
+  getStateSummary(): ClusterManagerStateSummary {
+    return {
+      maxAssignedNodes: this.state.maxAssignedNodes,
+      assignedNodes: this.getAssignedNodes().size,
+    };
+  }
+
+  getUnassignedNodes(): Set<string> {
+    return new Set(Array.from(this.state.nodes.keys()).filter((key) => this.state.nodes.get(key) === null));
+  }
+
+  getAssignedNodes(jobName?: string): Set<string> {
+    return new Set(
+      Array.from(this.state.nodes.keys()).filter((key) => {
+        const value = this.state.nodes.get(key);
+        return jobName ? value === jobName : value !== null && value !== 'BAD!';
```

Comment:
What's the expected behavior if `jobName` is an empty string? `Boolean("")` is `false`.